### PR TITLE
CU-2bbgg1k - Configurable constant product

### PR DESCRIPTION
**Pull request by hussein-aitlahcen**
## Issue
[Clickup issue](https://app.clickup.com/t/2bbgg1k)

## Description
Allow users to configure constant product pool weights. Prior to this PR, constant product pool were hardcoded with 50/50 weights. This PR allow arbitrary weights for base/quote assets. With the following property:
$$\sum_{i=1}^n{w_i}=1$$

As we have $n=2$, the implementation is expecting $w_1$ with $w_1 != 1$ to be provided and compute the according $w_2=1-w_1$.

Note: just wanted to try mathjax

## Checklist

- [x] I have updated the cargo docs to reflect changes made by this PR _(if applicable)_
- [x] I have updated the `book/` to reflect changes made by this PR _(if applicable)_

### CODE REVIEW

**Comment by ComposableFiB0t** [comment]
Task linked: [CU-2bbgg1k Create Pool Module - constant weights addition](https://app.clickup.com/t/2bbgg1k)

**Review by vivekvpandya — APPROVED** [review]

**Comment by KaiserKarel** [comment]
@hussein-aitlahcen needs rebase

**Review by dzmitry-lahoda — APPROVED** [review]

File: frame/composable-traits/src/dex.rs [hunk 1]
```diff
@@ -152,6 +152,10 @@ pub struct ConstantProductPoolInfo<AccountId, AssetId> {
 	pub fee: Permill,
 	/// Amount of the fee goes to owner of the pool
 	pub owner_fee: Permill,
+	/// The weight of the base asset. Must hold `1 = base_weight + quote_weight`
+	pub base_weight: Permill,
```
Comment (vimukthi-git) [inline frame/composable-traits/src/dex.rs:156]: Wouldn't it be more forward compatible if we define this as an array of weights? Given that we want to support multi asset eventually? Obviously we can reject the case with more than 2 for now

**Review by vimukthi-git — APPROVED** [review]
LGTM, one minor question though

File: frame/composable-traits/src/dex.rs (hunk 1 above)
Comment (hussein-aitlahcen) [reply frame/composable-traits/src/dex.rs:237]: If we do so we should do it for add/remove liquidity as well and integrate a vec everywhere

**Review by vimukthi-git — APPROVED** [review]
LGTM

File: frame/composable-traits/src/dex.rs (hunk 1 above)
Comment (KaiserKarel) [reply frame/composable-traits/src/dex.rs:237]: Note that the refactor will make a storage migration more difficult; I'd already refactor this to a Vec<Permill>, and check that len == 2; that way we can remove the check to allow multiweighted pools

**Review by KaiserKarel — APPROVED** [review]

**Pull request merged**